pkg/forward/forward.go

```diff
@@ -416,6 +435,91 @@ func (isdf *InterStepDataForward) forwardAChunk(ctx context.Context) {
 	forwardAChunkProcessingTime.With(map[string]string{metrics.LabelVertex: isdf.vertexName, metrics.LabelPipeline: isdf.pipelineName, "buffer": isdf.fromBuffer.GetName()}).Observe(float64(time.Since(start).Microseconds()))
 }
 
+// streamMessage streams the data messages to the next step. Return a bool to
+// indicate if the processing failed and the caller (forwardAChunk) needs to
+// return early.
+func (isdf *InterStepDataForward) streamMessage(
+	ctx context.Context,
+	dataMessages []*isb.ReadMessage,
+	processorWM wmb.Watermark,
+	readOffsets []isb.Offset,
+	writeOffsets *map[string][]isb.Offset,
+	messageToStep map[string][]isb.Message,
+) bool {
+	if isdf.opts.vertexType == dfv1.VertexTypeSource {
+		isdf.opts.logger.Errorw("Stream is not supported for source data transformer")
+		return true
+	}
+
+	// send to UDF only the data messages
+	if len(dataMessages) != 0 {
+		for _, m := range dataMessages {
+			// emit message size metric
+			readBytesCount.With(map[string]string{metrics.LabelVertex: isdf.vertexName, metrics.LabelPipeline: isdf.pipelineName, "buffer": isdf.fromBuffer.GetName()}).Add(float64(len(m.Payload)))
+			// assign watermark to the message. assign time.UnixMilli(-1) as watermark when we are at source vertex.
+			m.Watermark = time.Time(processorWM)
+
+			// process the UDF and get the result
+			start := time.Now()
+			udfReadMessagesCount.With(map[string]string{metrics.LabelVertex: isdf.vertexName, metrics.LabelPipeline: isdf.pipelineName, "buffer": isdf.fromBuffer.GetName()}).Inc()
+
+			writeMessageCh := make(chan isb.WriteMessage)
+			errs, ctx := errgroup.WithContext(ctx)
+			errs.Go(func() error {
+				return isdf.UDF.ApplyMapStream(ctx, m, writeMessageCh)
+			})
+
+			// Stream the message to the next vertex. First figure out which vertex
+			// to send the result to. Then update the toBuffer(s) with writeMessage.
+			for writeMessage := range writeMessageCh {
+				// if we do not get a time from UDF, we set it to the time from (N-1)th vertex
+				if writeMessage.EventTime.IsZero() {
```

Comment:
The eventTime is not supposed to be changed in the map udf, it's already specified in pkg/udf/function/uds_grpc.go, line 178.

Reply:
This is a replication of non streaming case at L695. Do you mean it is not needed? Then we are doing it for non streaming case?

Reply:
Technically it's not needed I think.